### PR TITLE
rfc: task list with json, uuid and duration

### PR DESCRIPTION
**Pull request by ekeren**
This is a proposal for story 13, that includes: 
- Json and MutJson
- Using `bring jsii` in `inflight` (replace counter with uuid.v4())
- Using duration
- Bucket cross region replication 
- VScode autocomplete capabilities 
- Least privilege permissions 

@yoav-steinberg, please notice that we are using a bunch of static functions in this code, mainly for 
str<->Json parsing and writing. Can you suggest an alternative api for static functions in this case? 

*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+- Least privileged permissions are granted
+- The code should work on sim, aws, azure
+- VScode should be able to autocomplete 
+- Console should be able to show all resources, and interact with them
+
+
+
+```ts (wing)
+bring cloud;
+// need to run: npm install uuid --save
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:23]: I think we will need a jsii wrapper for uuid in order for this to work.

```suggestion
// need to run: npm install @winglibs/uuid --save
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  init() {
+    this._bucket = new cloud.Bucket();
+  }
+
+  /** 
+   * Adds a task to the task list.
+   * @returns The ID of the new task.
+   */
+  inflight add_task(title: str): str {
+    let id = uuid.v4();
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:42]: This would be something like a static method because we don't have free floating functions.

```suggestion
    let id = uuid.Uuid.v4();
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    let j = Json { 
+      title: title, 
+    };
+    print("adding task ${id} with data: ${j}"); //j should be printed out nicely 
+    this._bucket.put(id, j.to_str());
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:47]:
```suggestion
    this._bucket.put_json(id, j);
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    return id;
+  }
+
+  /** 
+   * Gets a task from the task list.
+   * @param id - the id of the task to return
+   * @returns the title of the task (optimistic)
+   */
+  inflight get_task(id: str): Json {
+    let blob = this._bucket.get(id);
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:57]:
```suggestion
    let blob = this._bucket.get_json(id);
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  /** 
+   * Sets effort estimation on a test
+   * @param id - the id of the task to return
+   * @param effort_estimation - the time (duration) estimated for this task
+   * @returns The ID of the existing task.
+   */
+  inflight add_estimation(id: str, effort_estimation: duration) {
+    let j = this.get_task(id).to_mut();
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:68]:
```suggestion
    let j = this.get_task(id).clone_mut();
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    j.set("effort_estimation", effort_estimation);
+    this._bucket.put(id, j.to_str());
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:70]:
```suggestion
    this._bucket.put_json(id, j);
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+   /** 
+    * Gets the tasks ids 
+    * @returns set of task id
+    */
+  inflight list_task_ids(): Set<str> {
+    let result = MutSet<str> {};
+    for id in this._bucket.list() {
+      result.add(id);
+    }
+    return result.to_immut();
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:93]:
```suggestion
    return result.clone();
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      let j = this.get_task(id); 
+      let title = Str.from_json(j.get("title"));
+      if title.contains(term) { 
+        print("found task ${id} with title \"${title}\" with term \"${term}\"");
+        output.push(id);
+      }
+    }
+    
+    print("found ${output.len} tasks which match term '${term}'");
+    return output.to_immut();
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:116]:
```suggestion
    return output.clone();
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+  inflight before() {
+    for id in this.tasks.list_task_ids() {
+      this.tasks.remove_tasks(id);
+    }
+  }
+
+  init(name: str, test: inflight (Test): void) {
+    this.tasks = new TaskList();
+    this._test = test;
+    new cloud.Function({
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:137]: This really feels like a (very nice) hack... let's think how we can make this a little nicer...

Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:137]: We are still hacking this for now. It turns out that this hack will not work for this sprint... 

I'll have to change this for now (probably tomorrow when we know what is actually in)

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 7 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:93]: I find it counter intuitive that clone returns a different object type. 
I would expect reserving the `clone` for 🤷  cloning

Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:93]: BTW, I also don't like the `to_immut()` API 😊

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  /** 
+   * Gets a task from the task list.
+   * @param id - the id of the task to return
+   * @returns the title of the task (optimistic)
+   */
+  inflight get_task(id: str): Json {
+    let blob = this._bucket.get(id);
+    return Json.parse(blob)
```
Comment (staycoolcall911) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:58]:
```suggestion
    return Json.parse(blob);
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  /** 
+   * Sets effort estimation on a test
+   * @param id - the id of the task to return
+   * @param effort_estimation - the time (duration) estimated for this task
+   * @returns The ID of the existing task.
+   */
+  inflight add_estimation(id: str, effort_estimation: duration) {
```
Comment (staycoolcall911) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:67]:
```suggestion
  inflight add_estimation(id: str, effort_estimation: duration): str {
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 4 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:57]: I don't expect `str` to be a subtype of `Json` 

```ts (wing)
print_json(j: Json) {
   if (j.some_bool_value) {
      print(j.some_str_value)
   }
}

print_json("I am not a json");
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 1 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:23]: Life is tuff

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 3 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:47]: What's the problem?

Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:47]: I believe that there is a contractual difference between the two API:
1. `this._bucket.put(id, j.to_str());` - We are saving the string representation of a json
1. `this._bucket.put_json(id, j);` - We are saving a json (it doesn't say how it is saved)

Personally, I prefer the  number 1, if someone wants to use a Bucket for storing json's, he can create his own JsonBucketStore. If someone wants to create his own Bucket for storing parquet files or csv files he can creates his own abstractions on top of the regular bucket

Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:47]: But think goes back to me believing that we should have a thin API for our SDK, which may contradicts with "always optimize for the common case" in some cases

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 7 above)
Comment (MarkMcCulloh) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:93]: I'm a fan of clone or copy. Maybe copy is better because there's less implication of "deep" copy (which is not what we're doing). It's also 1 fewer letter :)

`result.copy()` and `result.copy_mut()` are nice because the effect is clear, the fact that the type changes doesn't feel as important to put in the name because it's already in the signature.

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 1 above)
Comment (MarkMcCulloh) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:23]: is uuid stuff important enough to include in the stdlib?

**Comment by ekeren** [comment]
for myself, need to add a definition and adjust the code for permissions

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 4 above)
Comment (Chriscbr) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:57]: Ok, maybe I jumped the gun here -- so even though `x` in my example above could be a string-ish value, maybe it's not correct to say `Json "hi"` is the same as `"hi"`.

Eyal raised this question ("if string is a subtype of json, then is this allowed?") that stumped me...

```
"hello".some_bool_value
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+> **Status**: Expected released on 2023/02/16
+
+The following code is an iteration of Wing's task list.
+This version of it includes the following functional changes:
+- Using npm's uuid package inflight instead of the counter resource 
+- Storing Json instead of a string
+- Using duration for estimated effort
+
+Also, it includes some non-functional requirements: 
+- The Bucket should be replicated across different regions
```
Comment (hasanaburayyan) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:12]: Should the story indicate that this is toggled featured rather than change in implementation of underlying resources?

Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:12]: I think this should be a plugin and the story should be explicit about how this plugin is applied.

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+- Storing Json instead of a string
+- Using duration for estimated effort
+
+Also, it includes some non-functional requirements: 
+- The Bucket should be replicated across different regions 
+  - +100 if we are able to do the same for azure as well
+- Least privileged permissions are granted
+- The code should work on sim, aws, azure
+- VScode should be able to autocomplete 
+- Console should be able to show all resources, and interact with them
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:17]: Let's add "Windows support" (we want to mitigate the risk)

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+The following code is an iteration of Wing's task list.
+This version of it includes the following functional changes:
+- Using npm's uuid package inflight instead of the counter resource 
+- Storing Json instead of a string
+- Using duration for estimated effort
+
+Also, it includes some non-functional requirements: 
+- The Bucket should be replicated across different regions 
+  - +100 if we are able to do the same for azure as well
+- Least privileged permissions are granted
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:14]: I think we should punt this to the next story

File: docs/06-contributors/999-rfcs/000-stories/story-13.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+- The code should work on sim, aws, azure
+- VScode should be able to autocomplete
```
Comment (eladb) [inline docs/06-contributors/999-rfcs/000-stories/story-13.md:16]: 👍 👍

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 1 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:23]: I actually think it is important enough to include in the stdlib at some point, but for the sake of this story, @ekeren wanted some simple external library, and this is good enough I believe.

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 3 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:47]: We don't have a design tenet to create a "thin API". I believe in a rich, "batteries included", SDK that will make people's lives as easy as possible. If we know there's a common case to put JSON objects into buckets, we should support `put_json()` and not make every person in the world implement it themselves.

> if someone wants to use a Bucket for storing json's, he can create his own JsonBucketStore

This is exactly what I want to avoid :-)

Also, think of `put_json()` simply as an overload (we don't have method overloads at the moment, so this is the best we can do).

We should definitely talk about this more generally from a design tenets perspective, and not in this sprint.

I think a good place to have the conversation about JSON is over https://github.com/winglang/wing/issues/1053 and maybe we should expand our SDK design tenets to cover this question about "minimal" versus "batteries included" API and how we offer "overloads" in our APIs (<-- @Chriscbr).

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 4 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:57]: I don't think `bucket.put()` should always take a JSON because there are definitely use cases to store objects that are not JSON (e.g. binary blobs).

Maybe in the future we should define a "Serializable" interface ("trait") and have both `str` and `Json` (and `Blob`) implement it, and then `put()` just takes a `Serializable`.

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 7 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:93]: We are having two separate discussions here:

1. Should we call this method `copy` or `clone`?
2. What should be the name of the method that returns an immutable copy and a mutable one?

The rationale of having `copy`/`clone` return an immutable version is because we want Wing to be ["immutable by default"](https://docs.winglang.io/blog/2023/02/02/good-cognitive-friction). So I am actually comfortable with the fact that this is counter intuitive (this is what I was calling "good cognitive friction").

I am less comfortable with `clone()` returning a *mutable* copy because this means that users can mistakingly use a mutable object and we want that to be a very explicit thing.

So in my view, the alternative to `copy()`+`copy_mut()` are `copy_immut()`+`copy_mut()` (that is, no `clone()`), which TBH, is quite ugly.

Bottom line, I still think `clone()` (or `copy()`) should return the immutable version, and if the user is "surprised" and has to explicitly request the mutable copy, I am happy about it.

(I am less opinionated about `clone` versus `copy` :-))

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 9 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:137]: Why do we need this actually? Why not just a bunch of tests?

**Review by eladb — DISMISSED** [review]

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 4 above)
Comment (Chriscbr) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:57]: @eladb I totally agree - I don't think only accepting `Json` should be our endgame. My original comment was just trying to point out that if our choices for story 13 are between the following:

```
// API A
resource Bucket {
  ...
  inflight put(key: str, value: str) { ... }
  inflight put_json(key: str, value: Json) { ... }
}
```

```
// API B
resource Bucket {
  ...
  inflight put(key: str, value: Json) { ... }
}
```

Then API B might be cleaner and less confusing for the beta / pre-beta.

Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:57]: But then how would you store a string in the bucket?

`"hello".to_json()` returns `"hello"` (with quotes).

How about we do API A and rename `put()` to `put_str()`.

(basically `put_json()` and `put_str()` and `get_json()` and `get_str()`).

Comment (Chriscbr) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:57]: Ahh, gotcha - `"hello"` can be parsed back into JSON while `hello` cannot. Not sure why I got so tripped up on this... yeah separate methods for these make sense to me now.

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 14 above)
Comment (staycoolcall911) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:14]: This is the next thing @yoav-steignberg will be working on. In sprint 13 we're not expecting a full solution for "permissions", just some progress.

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 13 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:17]: Windows support is only for the wing cli. 
Once the cli works on windows,  @ainvoner will do a spike to see what is missing in the console

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 1 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:23]: Do we want to use the`@winglibs` namespace  for this?

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 7 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:93]:
```suggestion
    return result.copy_mut();
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 14 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:14]: yes, this will be discussed in further details on the actual ticket

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 8 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:116]:
```suggestion
    return output.copy();
```

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 9 above)
Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md:137]: moving them to simple `cloud.Function` tests

Comment (ekeren) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md]: done

File: docs/06-contributors/999-rfcs/000-stories/story-13.md (hunk 1 above)
Comment (eladb) [reply docs/06-contributors/999-rfcs/000-stories/story-13.md]: We don't have to. We can just publish something like wing-uuid for now

**Review by staycoolcall911 — APPROVED** [review]
@ekeren - I think we can move forward with this, so I'm approving. Let's see if we need to limit the scope for the next sprint tick and or create an additional stories for bring jsii + permisssions
